src/main/scala/com/fulcrumgenomics/commons/async/AsyncIterator.scala

```diff
@@ -0,0 +1,94 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package com.fulcrumgenomics.commons.async
+
+import java.util.concurrent.{ArrayBlockingQueue, BlockingQueue, LinkedBlockingDeque}
+
+object AsyncIterator {
+  /** Builds an [[AsyncIterator]] and starts it.
+    *
+    * @param source the source iterator to consume
+    * @param bufferSize the maximum number of elements to buffer before the blocking when consuming the source, or None
+    *                   if unbounded.
+    * @tparam T the type of object to consume
+    */
+  def apply[T](source: Iterator[T], bufferSize: Option[Int] = None): AsyncIterator[T] = {
+    new AsyncIterator[T](source, bufferSize).start()
+  }
+}
+
+/** An asynchronous wrapper for an [[Iterator]].  A separate thread will be created to consume the source iterator.
+  * Will buffer up to [[bufferSize]] elements before the blocking when consuming the source.
+  *
+  * @param source the source iterator to consume
+  * @param bufferSize the maximum number of elements to buffer before the blocking when consuming the source, or None
+  *                   if unbounded.
+  * @tparam T the type of object to consume
+  */
+class AsyncIterator[T](private val source: Iterator[T], bufferSize: Option[Int] = None) extends Iterator[T] with AsyncRunnable {
+  bufferSize.foreach(size => require(size > 0, s"bufferSize must be greater than zero when given, found $size"))
+
+  private val queue: BlockingQueue[T] = bufferSize match {
+    case Some(size) => new ArrayBlockingQueue[T](size)
+    case None       => new LinkedBlockingDeque[T]()
+  }
+
+  protected def execute(): Unit = this.source.foreach(this.queue.put)
+
+  /** Returns true if there exists more elements, false otherwise */
+  def hasNext(): Boolean = {
+    checkAndRaise()
+
+    // If the queue is empty but the underlying thread is not done, then there may or may not be new
+    // items that will be put in the queue.
+    while (this.queue.isEmpty && !this.done) {
```

Comment:
I think this could be done as the following if you cached the next item like in a buffered iterator:

```scala
var item = null
while (item == null && !this.done) item = this.queue.poll(50, TimeUnit.Milliseconds)
...
```

This has the advantage that instead of always waiting 50ms when the queue is empty and the thread is working, it will wait _at most_ 50ms, but should return much sooner if an item is inserted into the queue.

Reply:
Done